bulk-model-sync-mps/src/main/kotlin/org/modelix/mps/model/sync/bulk/MPSBulkSynchronizer.kt

```diff
@@ -141,6 +151,118 @@ object MPSBulkSynchronizer {
+        access.executeCommandInEDT {
+            if (baseVersion != null) {
+                val invalidationTree = InvalidationTree(1_000_000)
+                val newTree = version.getTree()
+                fun ancestorsAsArray(nodeId: Long) = newTree.ancestorsAndSelf(nodeId).toList().asReversed().toLongArray()
+                newTree.visitChanges(baseVersion.getTree(), object : ITreeChangeVisitorEx {
+                    override fun containmentChanged(nodeId: Long) {
+                        // Containment can only change if also the children of the parent changed.
+                        // Synchronizing the parent will automatically update the containment of the children.
+                    }
+
+                    override fun childrenChanged(nodeId: Long, role: String?) {
+                        invalidationTree.invalidate(ancestorsAsArray(nodeId))
+                    }
+
+                    override fun referenceChanged(nodeId: Long, role: String) {
+                        invalidationTree.invalidate(ancestorsAsArray(nodeId))
+                    }
+
+                    override fun propertyChanged(nodeId: Long, role: String) {
+                        invalidationTree.invalidate(ancestorsAsArray(nodeId))
+                    }
+
+                    override fun nodeRemoved(nodeId: Long) {
+                        // same reason as in containmentChanged
+                    }
+
+                    override fun nodeAdded(nodeId: Long) {
+                        invalidationTree.invalidate(ancestorsAsArray(nodeId))
+                    }
+                })
+                val treePointer = TreePointer(newTree)
+                treePointer.runRead {
+                    ModelSynchronizer(
+                        invalidationTree,
+                        treePointer.getRootNode(),
+                        MPSRepositoryAsNode(repository),
+                        NodeAssociationFromMPS2Server(true, MPSArea(repository), treePointer)
+                    ).synchronize()
+                }
+            }
+
+            val allModules = repository.modules
+            val includedModules: Iterable<SModule> = allModules.filter {
+                isModuleIncluded(it.moduleName!!, includedModuleNames, includedModulePrefixes)
+            }
+            val numIncludedModules = includedModules.count()
+            val repoAsNode = MPSRepositoryAsNode(repository)
+            println("Importing modules...")
+            try {
+                println("Importing modules...")
+                // `modulesToImport` lazily produces modules to import
+                // so that loaded model data can be garbage collected.
+                val modulesToImport = includedModules.asSequence().flatMapIndexed { index, module ->
+                    println("Importing module ${index + 1} of $numIncludedModules: '${module.moduleName}'")
+                    val fileName = inputPath + File.separator + module.moduleName + ".json"
```

Comment:
Deriving the filename is also part of the Gradle plugin code. Do we have a shared library where that function could go to to avoid duplication and potential discrepancies later on?